GridDevice serialization refactor

verult opened the pull request:
This refactor adds the `to_proto()` method to serialize `GridDevice` into `DeviceSpecification`. The convenience method to create a `DeviceSpecification` from partial gateset information, `_create_device_specification_proto()`, is also reimplemented as `GridDevice.from_device_information()` and returns a `GridDevice` instead. These features makes using server-side GridDevice viable - Server-side and client-side GridDevice instances should now be identical.

In addition, gateset serialization and deserialization mappings are reorganized into a shared data structure.

cc @maffoo @pavoljuhas

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -74,6 +68,80 @@
+    """GateFamilies which match all other valid gate representations."""
+    additional_forms: List[cirq.GateFamily] = field(default_factory=list)
+    """`primary_forms` (as GateFamilies) + `additional_forms`"""
+    all_forms: List[cirq.GateFamily] = field(init=False)
```
Optional:  Should this be a generated @property instead?  I guess, with init=False, they can't specify it, so maybe that's okay.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -74,6 +68,80 @@
+    """The name of gate type in `GateSpecification`."""
+    gate_spec_name: str
+    """Gate representations to be included in generated gatesets and gate durations."""
+    primary_forms: List[GateOrFamily]
```
The primary and additional forms definition is kind of confusing to me.  This new terminology is not explained very clearly by these docstrings.  Maybe the class docstring could give more context as to what's going on here?

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+            DeviceSpecification.
+        """
+        qubits = self._metadata.qubit_set
+        unordered_pairs = [tuple(pair_set) for pair_set in self._metadata.qubit_pairs]
```
Should we sort self._metadata.qubit_pairs so that the order is stable.  I think qubit_pairs is just defined as an Iterable, so I am not sure if the ordering would be stable.

verult replied:
True, added a sort to `pairs` below.

dstrain115 commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+        gate_duration information: for every distinct gate, only one representation needs to be in
+        gateset and gate_duration. The remaining representations will be automatically generated.
+
+        For example, if the input gateset contains only `cirq.PhasedXZGate`, and the input
```
Do you have a test for this expansion of compatible gatesets?

verult replied:
It's tested in [`test_device_from_device_information_equals_device_from_proto`](https://github.com/quantumlib/Cirq/blob/4837660ed7d9aa2d415bd632c381abd5926d1e85/cirq-google/cirq_google/devices/grid_device_test.py#L473) by checking that the generated device matches the one from `from_proto()`. The gateset in the device from `from_proto()` is verified in [`test_grid_device_from_proto`](https://github.com/quantumlib/Cirq/blob/4837660ed7d9aa2d415bd632c381abd5926d1e85/cirq-google/cirq_google/devices/grid_device_test.py#LL278C46-L278C46).

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -74,6 +68,80 @@
+class _GateRepresentations:
+    """Contains equivalent representations of a gate in both DeviceSpecification and GridDevice."""
+
+    """The name of gate type in `GateSpecification`."""
+    gate_spec_name: str
+    """Gate representations to be included in generated gatesets and gate durations."""
+    primary_forms: List[GateOrFamily]
+    """GateFamilies which match all other valid gate representations."""
+    additional_forms: List[cirq.GateFamily] = field(default_factory=list)
+    """`primary_forms` (as GateFamilies) + `additional_forms`"""
+    all_forms: List[cirq.GateFamily] = field(init=False)
```
nit: Might be cleaner to put these comments in an `Attributes` section in the docstring:
```python
@dataclass
class _GateRepresentations:
    """Contains equivalent representations of a gate in both DeviceSpecification and GridDevice.

    Attributes:
        gate_spec_name: The name of gate type in `GateSpecification`.
        primary forms: Gate representations to be included in generated gatesets and gate durations.
        ...
    """

    gate_spec_name: str
    primary_forms: List[GateOrFamily]
    ...
```

verult replied:
Done

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+def _serialize_gateset_and_gate_durations(
+    out: v2.device_pb2.DeviceSpecification,
+    gateset: cirq.Gateset,
+    gate_durations: Mapping['cirq.GateFamily', 'cirq.Duration'],
```
nit: don't need to quote `cirq` types since we're in `cirq_google`:
```suggestion
    gate_durations: Mapping[cirq.GateFamily, cirq.Duration],
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+        getattr(gate_spec, gate_rep.gate_spec_name).SetInParent()
+
+        # Set gate duration
+        is_duration_set = False
+        for gf in gate_rep.all_forms:
+            if gf not in gate_durations:
+                continue
+
+            gate_duration_picos = int(gate_durations[gf].total_picos())
+            if is_duration_set and gate_duration_picos != gate_spec.gate_duration_picos:
+                raise ValueError(
+                    'Multiple gate families in the following list exist in the gate duration dict,'
+                    f' and they are expected to have the same duration value: {gate_rep.all_forms}'
+                )
+            is_duration_set = True
+            gate_spec.gate_duration_picos = gate_duration_picos
```
Could use a set comprehension to get all durations, then check the size afterward:
```suggestion
        gate_durations_picos = {
            int(gate_durations[gf].total_picos())
            for gf in gate_rep.all_forms
            if gf in gate_durations
        }
        if len(gate_durations_picos) > 1:
            raise ValueError(
                'Multiple gate families in the following list exist in the gate duration dict,'
                f' and they are expected to have the same duration value: {gate_rep.all_forms}'
            )
        elif len(gate_durations_picos) == 1:
            gate_spec.gate_duration_picos = gate_durations_picos.pop()
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+) -> v2.device_pb2.DeviceSpecification:
+    """Serializes the given gateset and gate durations to DeviceSpecification."""
+
+    gate_specs = {}
```
nit: add type annotation

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+        gate_specs[gate_name] = gate_spec
+
+    # Sort by gate name to keep valid_gates stable.
+    out.valid_gates.extend([v for _, v in sorted(gate_specs.items(), key=lambda item: item[0])])
```
nit: can omit brackets since `extend` can accept an iterable. Also, I think the explicit key function is not needed because dict keys are unique so the tuple comparison will only use the first element
```suggestion
    out.valid_gates.extend(v for _, v in sorted(gate_specs.items()))
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -120,41 +187,64 @@ def _validate_device_specification(proto: v2.device_pb2.DeviceSpecification) ->
+
+        # GateSpecification dedup. Multiple gates or GateFamilies in the gateset could map to the
+        # same GateSpecification.
+        gate_name = gate_spec.WhichOneof('gate')
```
Isn't this equivalent to the `gate_spec_name` which was set above? Could perhaps simplify by bringing that down here:
```python
gate_name = gate_rep.gate_spec_name
getattr(gate_spec, gate_name).SetInParent()
gate_specs[gate_name] = gate_spec
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+        # TODO(#5050) remove empty checks below once deprecated fields in DeviceSpecification are
+        # removed.
+
+        if len(out.valid_qubits) == 0:
```
nit: style guide says use false-y empty lists (https://google.github.io/styleguide/pyguide.html#2144-decision)
```suggestion
        if not out.valid_qubits:
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+
+        if len(out.valid_qubits) == 0:
+            known_devices.populate_qubits_in_device_proto(qubits, out)
+        if len(out.valid_targets) == 0:
```
here, too
```suggestion
        if not out.valid_targets):
```

maffoo commented on cirq-google/cirq_google/devices/grid_device.py:
```diff
@@ -373,6 +461,103 @@ def from_proto(cls, proto: v2.device_pb2.DeviceSpecification) -> 'GridDevice':
+            ValueError: If multiple gate families in gate_durations can
+                represent a particular gate, but they have different durations.
+        """
+        metadata = cirq.GridDeviceMetadata(
+            qubit_pairs=qubit_pairs, gateset=gateset, gate_durations=gate_durations
+        )
+        incomplete_device = GridDevice(metadata)
+        return GridDevice.from_proto(incomplete_device.to_proto())
```
This feels odd to create a device from metadata, then turn it into a proto and then back into a device. Is there a reason for the roundtrip through protos here? If so a comment would be helpful.

verult replied:
Added - lmk if the rationale doesn't make sense

verult reviewed (commented):
First pass at address comments.

verult commented:
In the latest push:
* Fixed coverage check
* Made _from_device_information to (1) discourage end-users from using it, and (2) make it easier to change in the future
* Renamed `_GateRepresentation` fields to better reflect their purposes.
